rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,124 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some possible drawbacks:
+
+- A new build system would require changes on the code
+  + On the other hand, such changes are likely to improve the code base.
+  
+- A new build system requires the developers become familiarized with it
+  - Specially when this build system uses its own description language.
+  + However, Meson is well documented, and its language is easy to grasp, specially for those familiarized with Python.
+
+- A new build system indirectly brings its own dependencies to the Nix project
+  - In particular, the reference implementation of Meson is written in Python.
+  - Further, this reference implementation generates script files meant to be consumed by Ninja, a tool written in C++ that acts like a Make replacement.
+  - This particular setting brings concerns about complexifying the bootstrap route.
+  + Given that Nix is currently written in C++, we can assume a C++ compiler as part of such a bootstrap route.
+  + There are full-featured alternative tools that replace Meson and Ninja. 
+    + Namely, Muon and Samurai are implementations of Meson and Ninja that require only a C compiler and a set of POSIX standard tools.
+
+- A new build system would require new strategies from the end users
+  - In particular, package managers that deploy Nix for their respective platforms.
+  + However, Meson and Ninja are nowadays a widespread toolset.
+    + Many open source projects use it, from mpv and dosbox-staging to Xorg and GNOME
+    + According to Repology, Meson is present in 53 package manager's families
+
+- The transition between between the old and new build systems should be smooth
+  - A wrapper script, maybe?
+  + Meson is not an obscure project; a mere documentation update should suffice
+
+# Alternatives
+[alternatives]: #alternatives
+
+The alternatives are
+
+- Doing nothing
+
+  It would keep the current code confusing and harder to work with.
+  
+- Other building systems (cmake, waf, premake etc.)
+  - Their strenghts and weaknesses should be evaluated.
+    - Tools strongly tied to other programming languages are strongly discouraged, because they further complexifies the bootstrap route as discussed above.
+      - Namely, waf is basically a Python library, whereas premake is a Lua library.
+    - Cmake has many noteworthy advantages:
+      + Can generates Make- and Ninja-compatible scripts;
+      + Supports Windows NT;
+      + Supports typical high level idiomatic constructions;
+      - On the other hand, the language is arguably more complex.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Questions that deserve further inquiry:
+
+- Unexpected interactions with Meson and Ninja
+  - Specially, vendoring and reproducibility.
+
+# Future work
+[future]: #future-work
+
+- Deprecate the current build scripts
+- Backport the new build system to Nix 2.3
```

Comment:
please add:

"- introduce a complimentry buildsystem using pure makefiles, or cmake to increase project flexibility and allow for redundancy. Many other open source projects use a multiple buildsystems to compliment each other."

Reply:
I am not a fan of such redundancies. I can't assure that we will have enough people to maintain two overlapping technologies.

That being said, I believe this is more suited to alternatives. Anyone is free to pick the ideas from there.

Reply:
i dont mind either way. buildsystem redundancy is common for major open source projects. allowing more flexibility with complimentary buildsystems can add resilience for the bootstrap-ability of the project. I think it would be smart to allow people the choice. also, maintainence can now be enforced with the CI platform that @Ericson2314 introduced.

Reply:
Buildsystem redundancy is relatively common for projects that provide a library, particularly as it eases integration within a project (for example, both Meson and CMake have ways to vendor a third-party library that uses the same buildsystem, by e.g. adding a Meson subproject, or by treating it as "yet another directory of your own source code" in cmake and just calling `add_subdirectory()`).

It's much less common for e.g. projects that provide *tools* instead of libraries. Simply providing multiple build systems "for the pleasure of people building it" isn't usually seen as an especially valuable goal.

Reply:
Of course, it's also common for projects that are switching build systems to keep both the old and the new one around for a release or two, to ensure that any kinks are worked out of the new one before tossing the one that people were previously using.

Reply:
mpv recently implemented Meson scripts as an option.

Reply:
Yes, and their long term plan is to get rid of waf. :D

(It is waf, can you blame them?)

Reply:
it isnt uncommon to offer multiple, especially if the community wants that